PKCS15 Emulator for Starcos 3.x Cards with eSign app

jozsefd opened the pull request:
##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [ ] Documentation is added or updated
- [x] New files have a LGPL 2.1 license statement
- [x] PKCS#11 module is tested
- [x] Windows minidriver is tested
- [ ] macOS tokend is tested

## Scope

New card types (SC_CARD_TYPE_STARCOS_V3_4_ESIGN and SC_CARD_TYPE_STARCOS_V3_5_ESIGN) and PKCS15 emulator for StarCOS 3.x cards with eSign app, released by Bundesagentur für Arbeit, a German Federal Agency. 
The card contains 2 2028-bit RSA keys and the associated certificates, intended for non qualified PKI operations. The card may include a QES app for qualified signatures, not included in this PR. 
The new card types are "emulation only" as the card profile may contain a broken PKCS#15 structure. The PKCS15 emulator supports only the user PIN, and by default, it emits the card holder certificates only. The compile time option ENABLE_ESIGN_ISSUER_CONTAINERS enables issuer certificates in the emulator.

## Notes

The StarCOS v3.5 card has no support for ISO7816 Pin-Info, it may prevent PIN operations in PKCS#11, e.g. in Firefox. The PR contains some modifications of the StartCOS card driver: among others, it includes PSS-Padding support, fixes Select by AID, fixes extended APDU handling and disables path-caching.

frankmorgner commented on src/libopensc/pkcs15-starcos-esign.c:
```diff
@@ -0,0 +1,364 @@
+
+	if (card->type != SC_CARD_TYPE_STARCOS_V3_4 && card->type != SC_CARD_TYPE_STARCOS_V3_5) {
+		return SC_ERROR_WRONG_CARD;
+	}
```
By nature, STARCOS is a chip OS that can be configured with a number of profiles. If someone had the time to implement this correctly, then parsing of EF_KEYD, EF_CERT, ... should be done. For sake of simplicity and due to encoding problems during card initialization, we often take shortcuts by providing a hard coded profile as done in [pkcs15-starcos-esign.c](https://github.com/OpenSC/OpenSC/pull/2544/files#diff-b8cf394129d75ae702ab117c61b6fd7d86761bf056b340844ac402dadf546057). However, I think you want to avoid matching generic STARCOS 3.5 cards (which may not exactly have the profile you are assuming here...)

Typically, OpenSC driver developers use a dedicated card type, say `SC_CARD_TYPE_STARCOS_V3_5_VDA`, which not only matches the STARCOS version, but also the specific ATRs of your cards. This way, you can use your emulator for your cards, but for other STARCOS 3.5 cards, it would still be possible to parse the default ISO7815-15 structure to search for keys, certificates and apps.

jozsefd replied:
Unfortunately the card profile has no unique ATR (at least for the StarCOS 3.4), I have a D-Trust card which responds with the same ATR, but with a completely different profile. That card has a correct EF_DIR and a proper #PKCS15 structure on card, and works directly with OpenSC. The cards with the eSign profile have several issues, some lacks the ISO #PKCS15 directory and some have a broken version. This emulator tries to fix these issues.
Anyway, I will double check if I can assign these cards a uniqe card type.

jozsefd replied:
I have double checked the ATRs of this profile.
StarCOS 3.5 eSign card seems to have a unique ATR, so a new card type SC_CARD_TYPE_STARCOS_V3_5_ESIGN could be defined. Unfortunately, in case of the v3.4 card, the profile uses a generic ATR, so AID checking is needed to determine if the card is SC_CARD_TYPE_STARCOS_V3_4_ESIGN.
I set these card types as emulation only.

jozsefd commented on src/libopensc/card-starcos.c:
```diff
@@ -339,18 +341,6 @@ static int starcos_init(sc_card_t *card)
-		if (card->ef_atr->max_command_apdu > 0) {
-			card->max_send_size = card->ef_atr->max_command_apdu;
-		}
-	}
```
I had issues with extended APDUs in starcos_decipher(): some cards have reported an invalid length in EF.ATR and causing an APDU formatting error.  I review this problem again and put back the parsing of EF.ATR if possible.

jozsefd replied:
I have reapplied the function sc_parse_ef_atr(). I had to adjust the reported maximum size values to prevent error response 67 00 (Wrong length) in iso7816_read_binary() or during decryption.
Some card profiles reported the length limits only in the IAS/ECC specific issuer tag, I added parsing of this value as well.

jozsefd commented on src/libopensc/pkcs15-starcos-esign.c:
```diff
@@ -0,0 +1,364 @@
+#include <string.h>
+
+/* compile time option: define ADD_ISSUER_CONTAINERS to enable issuer certificates */
+#undef ADD_ISSUER_CONTAINERS
```
Yes, you are right. What is the right place to document a compile time flag?

jozsefd replied:
Reviewed this compile time flag, renamed to a more explanatory ENABLE_ESIGN_ISSUER_CONTAINERS. I did not find a place for documentation, e.g. the flag FORCE_VERIFY_RUTOKEN appears only in the respective card driver source file.

Jakuje commented:
Please, rebase the PR on top of current master.

jozsefd commented:
> Please, rebase the PR on top of current master.

I used Sourcetree for rebase, hope it is OK.

jozsefd commented:
> It does not look like that. This PR now contains 100 commits (from master) and conflicts with current master.

Oops, I messed up, will try to fix it.

jozsefd commented:
> > It does not look like that. This PR now contains 100 commits (from master) and conflicts with current master.
> 
> Oops, I messed up, will try to fix it.

I have reviewed the changes and the possible options to fix this PR. The simplest solution seems to be dropping this PR and reapplying the changes in a new PR. 
During the discussion of this PR the scope was actually expanded: at the end, not only a PKCS15 emulator was added, but 2 new (synthetic only) card types were also defined. 

Do you agree with this solution?

jozsefd commented:
> I dont know how creating a new PR solves any issue. If you can fix the stuff in new PR in new branch, why it is not possible to do that in the same branch? `git fetch upstream && git rebase -i upstream/master && git push -f` should solve the issues (indeed after resolving potential conflicts or removing unneeded commits during the interactive rebase. What step from the above does not work?

Yes, you are right, creating a new PR would not solve any issue, but I got stuck with my Windows git tools and I also wanted a better description for the PR. In the meantime, I realized that the PR title and text is editable and after moving back to Linux, I could rebase the branch, as you suggested.

Jakuje commented on src/libopensc/card-starcos.c:
```diff
@@ -281,7 +292,64 @@ static unsigned int starcos_determine_pin_encoding(sc_card_t *card)
+	rv = starcos_select_mf(card);
+	if ( rv == SC_SUCCESS ) {
+		sc_file_t * file;
+		rv = starcos_select_aid(card, esign_aid, 9, &file);
```
please, try to avoid using magic numbers:
```suggestion
		rv = starcos_select_aid(card, esign_aid, sizeof esign_aid, &file);
```

Jakuje commented on src/libopensc/card-starcos.c:
```diff
@@ -1944,18 +2087,14 @@ static int starcos_get_serialnr(sc_card_t *card, sc_serial_number_t *serial)
 		return SC_SUCCESS;
 	}
 
-	switch (card->type) {
-		case SC_CARD_TYPE_STARCOS_V3_4:
-		case SC_CARD_TYPE_STARCOS_V3_5:
+	if ( IS_V3x(card) ) {
 			card->serialnr.len = SC_MAX_SERIALNR;
 			r = sc_parse_ef_gdo(card, card->serialnr.value, &card->serialnr.len, NULL, 0);
 			if (r < 0) {
 				card->serialnr.len = 0;
 				return r;
 			}
-			break;
-
-		default:
+	} else {
 			/* get serial number via GET CARD DATA */
 			sc_format_apdu(card, &apdu, SC_APDU_CASE_2_SHORT, 0xf6, 0x00, 0x00);
 			apdu.cla |= 0x80;
```
After conversion from `switch` to `if`, the block contents are indented too much

jozsefd replied:
I will fix it.

Jakuje commented:
The commit series is still very long and most of the commits are fixups and fixups of fixups (adding and removing lines back and forth). If they do not make sense on their own, please, squash them to self-contained commits

jozsefd commented:
> The commit series is still very long and most of the commits are fixups and fixups of fixups (adding and removing lines back and forth). If they do not make sense on their own, please, squash them to self-contained commits

I squashed the unnecessary commits, fixed the magic number and indenting issues.

Jakuje commented:
Code looks ok, but the CIFuzz found a new crash in your code. Can you check the report and fix it?
```
AddressSanitizer:DEADLYSIGNAL
=================================================================
==21==ERROR: AddressSanitizer: SEGV on unknown address 0x00067fff803f (pc 0x00000097f665 bp 0x7ffcd73d8da0 sp 0x7ffcd73d8d70 T0)
==21==The signal is caused by a READ memory access.
SCARINESS: 20 (wild-addr-read)
    #0 0x97f665 in sc_file_valid /src/opensc/src/libopensc/sc.c:821:15
    #1 0x97f287 in sc_file_free /src/opensc/src/libopensc/sc.c:630:23
    #2 0x61dff6 in starcos_has_esign_app /src/opensc/src/libopensc/card-starcos.c:350:3
    #3 0x60e4b0 in starcos_init /src/opensc/src/libopensc/card-starcos.c:461:50
    #4 0x574701 in sc_connect_card /src/opensc/src/libopensc/card.c:358:8
    #5 0x55fe1b in fuzz_connect_card /src/opensc/src/tests/fuzzing/fuzzer_reader.c:205:9
    #6 0x55d851 in LLVMFuzzerTestOneInput /src/opensc/src/tests/fuzzing/fuzz_card.c:60:6
    #7 0x457c43 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:611:15
2022-06-14 15:28:43,989 - root - INFO - Fuzzer: fuzz_card. Detected bug.
2022-06-14 15:28:43,990 - root - INFO - Trying to reproduce crash using: /tmp/tmpfqbc61tl/crash-da92fc3ae8a7d3793dfb92468aa1cc8132cace17.
    #8 0x45742a in fuzzer::Fuzzer::RunOne(unsigned char const*, unsigned long, bool, fuzzer::InputInfo*, bool, bool*) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:514:3
    #9 0x459294 in fuzzer::Fuzzer::ReadAndExecuteSeedCorpora(std::__Fuzzer::vector<fuzzer::SizedFile, std::__Fuzzer::allocator<fuzzer::SizedFile> >&) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:826:7
    #10 0x4594c9 in fuzzer::Fuzzer::Loop(std::__Fuzzer::vector<fuzzer::SizedFile, std::__Fuzzer::allocator<fuzzer::SizedFile> >&) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:857:3
    #11 0x448fff in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:912:6
    #12 0x471d12 in main /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerMain.cpp:20:10
    #13 0x7f87bd67d082 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x24082)
    #14 0x4220fd in _start (build-out/fuzz_card+0x4220fd)
DEDUP_TOKEN: sc_file_valid--sc_file_free--starcos_has_esign_app
AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: SEGV /src/opensc/src/libopensc/sc.c:821:15 in sc_file_valid
==21==ABORTING
MS: 0 ; base unit: 0000000000000000000000000000000000000000
0x80,0xff,0x3,0xd,0x95,0x49,0x49,0x49,0x49,0x49,0x49,0x3b,0xd0,0x97,0xff,0x81,0xb1,0xfe,0x45,0x1f,0x7,0x2b,
\200\377\003\015\225IIIIII;\320\227\377\201\261\376E\037\007+
```
https://github.com/OpenSC/OpenSC/runs/6883610649?check_suite_focus=true

jozsefd commented:
I managed to fix the CIFuzz crash.

Jakuje commented:
No further review comments so I will merge this change. Thank you for your contribution!
